### DISCUSSION TITLE
filter_through command for block selection

### DISCUSSION THREAD

**Discussion by smartmic** [orbitalquark/textadept#44] [General]
From API docs for `textadept.editing.filter_through(command)`

> If text is selected and spans multiple lines, all text **on the lines** that have text selected is passed as stdin. 

Would it be possible to change this behaviour in case of block selection so that only the selected rectangular block is passed to stdin? 

My use case which does not yet work in textadept: I have a text table with some numbers in several columns and want to select only one column in order to pass it to `awk` (or similar commands) to do some quick calculations on the column (like sum, mean, …). Right now this fails because the whole line will be sent as stdin despite a block selection. Block copy&paste already works as expected.

**Comment by orbitalquark** [comment]
It shouldn't be difficult to pass a block selection to a command over stdin. The problem is replacing the block selection's contents with the command's stdout. `buffer:replace_sel()` or `buffer:replace_target()` does not handle replacement of rectangular selections like block copy/paste, so I think code would have to be written especially for this case. It would have to handle cases where output does not have as many lines as input, columns in output not matching columns in input, and columns in output not being constant.

**Reply by smartmic** [reply]
Hm, I see. Just thinking loud …to insert text back from stdout, would it be possible to change into overtype mode and insert byte by byte from stdout, replacing newlines with a jump back to original cursor position and going one line down? Afterwards, I have overlayed the original text with stdout. To me, it is fine if the original text will not be replaced by whitespaces in the block selection (although that could be a different option).

**Comment by orbitalquark** [comment]
You are welcome to give it a try and let me know how it goes. The code is in *modules/textadept/editing.lua*.

**Reply by smartmic** [reply]
Okay, this is already a while back but I did come up with a solution. At least it works as expected from my end. The code is a little bit more complex to handle also situations where the output of the external command will span beyond the length of the buffer.

I added a new function [_overlay_text_](https://github.com/smartmic/textadept/blob/37e11546a3e14399cd9fc4d11b2f9bcfae3be5ba/modules/textadept/editing.lua#L598) as well as modified _M.filter_through_.

Let me know your thoughts, I would be happy to contribute back if it is interesting.

Here is a short screencast for demonstration showing 2 example use cases:

https://user-images.githubusercontent.com/6379140/120085078-83660380-c0d5-11eb-829e-70f9b7b2042c.mp4

1. I mark a column of numbers and run `awk` and [`datamash`](https://www.gnu.org/software/datamash/) to do some calculations on that array. To keep the selected numbers as is I use [`pee`](https://www.mankier.com/1/pee) `cat "<command>"`, otherwise the result of the calculation will overwrite the first "cell" only in case of a single number output.
2. I use `tac` to reverse the block of selected strings

**Comment by orbitalquark** [comment]
Cool, I'll take a look when I have some time.

**Comment by orbitalquark** [comment]
Okay, I took a look at this and ended up submitting an upstream patch to Scintilla (Textadept's editing component) to take advantage of an internal routine that handles pasting text into a rectangular selection, which is more or less what you are doing. If and when it lands, I'll change `textadept.editing.filter_through` to operate on both multiple selections and rectangular selections. In my testing so far, all looks good.

**Comment by orbitalquark** [comment]
The next nightly release will support multiple and rectangular selections: https://github.com/orbitalquark/textadept/commit/48134170e780091b753d11dd945547a719719dd3

**Reply by smartmic** [reply]
Thanks for your efforts for patching this upstream without custom Lua code! 

However, after my first tests, it does not yet work as expected though. Just an example. Given this table:
```
+─────+──────────+──────+─────────+
| id  | name     | age  | gender  |
+─────+──────────+──────+─────────+
| 1   | Roberta  | 39   | M       |
| 2   | Oliver   | 25   | M       |
| 3   | Shayna   | 18   | F       |
| 4   | Fechin   | 18   | M       |
+─────+──────────+──────+─────────+
```
when I make a rectangular selection  just around the numbers in column 'age' and pass this through the external command `tac`, I get:
```
+─────+──────────+──────+─────────+
| id  | name     | age  | gender  |
+─────+──────────+──────+─────────+
| 1   | Roberta  | 1818   | M       |
| 2   | Oliver   | 25   | M       |
| 3   | Shayna   | 39   | F       |
| 4   | Fechin   |    | M       |
+─────+──────────+──────+─────────+
```
Similar strange behaviour when chosing the first three chars of all names and trying to reverse selection with `tac`.

**Comment by orbitalquark** [comment]
A quick fix is to change the line https://github.com/orbitalquark/textadept/commit/48134170e780091b753d11dd945547a719719dd3#diff-398fb21ba27b77c217e9de359157850346fdbc5610a8e0515171e6eb71b04cf3R627

  inout = table.concat(inout, '\n') .. '\n'

(append a trailing newline)

I haven't fully tested or thought about the consequences, but that seems to work at least for 'tac'.

**Reply by smartmic** [reply]
Yeah, that works perfectly! Not only `tac` but also some quick test with calculations of numbers via `awk` or `datamash` were successful.

This is a great enhancement for the editor, thank you very much @orbitalquark . Textadept is really the hidden gem under all editors, I am so glad I switched :four_leaf_clover:

**Comment by orbitalquark** [comment]
Thanks for the feedback. I've committed this: https://github.com/orbitalquark/textadept/commit/d76d3040f7fbe2eed89ba0af1411b5950c0ef3ad
